proposals/text/4084-agent-demo.md

```diff
@@ -0,0 +1,191 @@
+- Title: Agent tools
+- Decision driver: Vladimir Blagojevic (with Julian Risch)
+- Start Date: 2023-02-06
+- Proposal PR: https://github.com/deepset-ai/haystack/pull/4084
+
+
+# Summary
+
+Agent system consists of an extendable set of modules (aka experts/tools) and a "router" dispatching natural
+language input to a module that can best respond to the given input. These modules can be:
+
+• Neural, including the general-purpose LLMs as well as other smaller, specialized LMs.
+
+• Symbolic, for example a math calculator, a currency converter, or an API call to a search engine.
+
+
+Current Haystack pipeline design approach fits nicely with these extension modules/tools. They can be tested in
+isolation and subsequently integrated into the agent. The agent can register these tools, understand their objectives
+and route the user input to the appropriate tool. The tools can be either Haystack pipelines or components.
+
+
+A demo idea for the new Agent is to demonstrate Question Answering on Technical Documentation (using the example of Haystack). It should be useful and impressive, and it determines what tools we will implement first.
+
+**Example Questions:**
+- "Why am I seeing duplicate answers being returned?" based on indexed FAQ documentation
+- "Which organizations use Haystack?" based on web search and Wikipedia returning answers via SerpAPI
+- "How can I choose the model for PromptNode?" based on retrieving documents via WebRetriever
+- "How can I make overwrite_with_env_variables work in RayPipeline" based on an open issue found with GitHub API or web search
+
+If an answer cannot be found in indexed files, the Agent will use self reflection to rephrase the question and/or search the web. It will give updates while searching, for example print thoughts: “Found nothing in indexed documentation. Will continue with web search.” If still nothing can be found, the Agent will generate a link to a pre-filled and pre-tagged issue template like this that the user can choose to create. Tools required for the demo: SerpAPI, GitHubAPI, Self Reflection Module, WebRetriever.
```

Comment:
Does the web search only happen with some max limit? or can it go into infinite loops?

Reply:
@mayankjobanputra A parameter `max_iterations` will prevent infinite loops. It's implemented in the Agent so the tools should not need to check that.